src/main/java/net/neoforged/neoforge/event/ValidityEvent.java

```diff
@@ -0,0 +1,134 @@
+import net.minecraft.world.entity.vehicle.ContainerEntity;
+import net.minecraft.world.level.Level;
+import net.minecraft.world.level.block.entity.BlockEntity;
+import net.neoforged.neoforge.common.NeoForge;
+import net.neoforged.neoforge.event.entity.player.PlayerEvent;
+import org.jetbrains.annotations.ApiStatus;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * Event to check if a menu can stay open or if the player has moved too far away from it.
+ * <p>
+ * This is intended for mods that allow a player remote access to blocks and their menus,
+ * but it also can be used to forcibly kick them out of menus.
+ * <p>
+ * You can query the result of the original distance check with {@link #getOriginallyValid()}
+ * and set your override with {@link #setStillValid(boolean)}.
+ * <p>
+ * This base class is never posted (but it can be subscribed to, as it is not abstract),
+ * only one of its subclasses:
+ * <ul>
+ * <li>{@link ValidityEvent.Menu} for menus that have a {@link BlockPos} and may or may not be container-based,
+ * <li>{@link ValidityEvent.ContainerEntityMenu} for {@link ContainerEntity}-based menus, or
+ * <li>{@link ValidityEvent.EntityMenu} for menus attached to an {@link Entity}, or
+ * <li>{@link ValidityEvent.BlockEntityMenu} for the checks done by signs which don't use menus but a Screen
+ * and modded menu-like Screens that opt-in to use this.
+ * </ul>
+ * This event is fired on the {@link NeoForge#EVENT_BUS}.
+ */
+public class ValidityEvent extends PlayerEvent {
+    private final double distance;
+    private final boolean originallyValid;
+    private Optional<Boolean> stillValid = Optional.empty();
+
+    ValidityEvent(Player player, double distance, boolean originallyValid) {
+        super(player);
+        this.distance = distance;
+        this.originallyValid = originallyValid;
+    }
+
+    public double getDistance() {
+        return distance;
+    }
+
+    public boolean getOriginallyValid() {
+        return originallyValid;
+    }
+
+    public void setStillValid(boolean result) {
+        this.stillValid = Optional.of(result);
+    }
+
+    public boolean getStillValid() {
+        return stillValid.orElse(originallyValid);
+    }
+
+    public static class BlockBased extends ValidityEvent {
+        private final Level level;
+        private final BlockPos blockPos;
+
+        BlockBased(Player player, Level level, BlockPos blockPos, double distance, boolean originallyValid) {
+            super(player, distance, originallyValid);
+            this.level = level;
+            this.blockPos = blockPos;
+        }
+
+        public Level getLevel() {
+            return level;
+        }
+
+        public BlockPos getBlockPos() {
+            return blockPos;
+        }
+    }
+
+    public static class Menu extends BlockBased {
+        @ApiStatus.Internal
+        public Menu(Player player, Level level, BlockPos blockPos, double distance, boolean originallyValid) {
```

Comment:
I'm not so sure on this being internal. If you're expecting the `stillValid` check to end up firing the event, a mod may have a container whose validity is not based on a distance, but on some other arbitrary value, meaning they'd need to fire the event themselves, and the `CommonHooks` method won't help them.

Reply:
Because the whole Event tree is only about the distance (I think I'll rename it. I've been avoiding the old name because of the license, but the new one just isn't good.), so its result should not overrule any non-distance-based checks. For this, there are 3 different things to calculate the distance to, BlockPos, AABB, and Entity. The first case has 2 sub-cases, one in which a menu is open and the hacky way signs use pure Screens to mimic what menus do. (A remnant from the time when menus required the block to have an inventory container.)

If an event subscriber only cares about the BlockPos and distance, they can subscribe to BlockBased. If a modded block wants to have a GUI, they better have a menu and not copy the whacky thing Sign does, so they can fire Menu.

All constructors are internal/package because there are entry points in CommonHooks for all cases.